src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java

```diff
@@ -526,4 +515,62 @@ public boolean isAvailable(final String component, final String key, final Local
         }
         return available;
     }
+
+    /**
+     * Retrieves the localized message
+     *
+     * @param locale The locale in which the message is requested to be localized
+     * @param component The Singleton component in which the message belongs
+     * @param key The key that represents the message
+     * @param args named arguments to replace placeholders in the message
+     * @return One of the items in the following priority-ordered list:.
+     * @throws VIPJavaClientException If none from the list below is available
+     * <ul>
+     * 		<li>The source message, if source message hasn't been collected and translated</li>
+     * 		<li>The message in the requested locale</li>
+     * 		<li>The message in the next available fallback locale</li>
+     * 		<li>The source message</li>
+     * </ul>
+     */
+    public String getMessage(final Locale locale, final String component, final String key, final Map<String, Object> args) {
+        // Use source message if the message hasn't been collected/translated
+        String source = getLocaleMessages(Locale.forLanguageTag(ConstantsKeys.SOURCE), component).get(key);
+        if (source!=null && !source.isEmpty()) {
+            String collectedSource = getLocaleMessages(LocaleUtility.getSourceLocale(), component).get(key);
+            if (!source.equals(collectedSource)) {
+                return FormatUtils.formatMsg(source, LocaleUtility.getSourceLocale(), args);
+            }
+        }
+
+        String message = null;
+        Locale localeInUse = locale;
+        Map<String, String> messages = getLocaleMessages(locale, component);
+        if (messages == null || messages.isEmpty()) {
+            Iterator<Locale> fallbackLocalesIter = LocaleUtility.getFallbackLocales().iterator();
+            while (fallbackLocalesIter.hasNext()) {
+                localeInUse = fallbackLocalesIter.next();
+                if (!localeInUse.equals(locale)) {
+                    messages = getLocaleMessages(localeInUse, component);
+                    if (messages != null && !messages.isEmpty()) {
+                        break;
+                    }
+                }
+            }
+        }
+
+        if (messages!=null) {
+            message = messages.get(key);
+        }
+        if (message == null || message.isEmpty()) {
+            throw new VIPJavaClientException(FormatUtils.format(ConstantsMsg.GET_MESSAGE_FAILED, key, component, locale));
+        }
```

Comment:
Because formatting needs to know the locale to use, if locale fallback is taken care of in the service layer, then formatting doesn't know the correct locale to use. it will use current locale always.

Reply:
You do not need to create another object called LocaleMessages.  Just add locale to MessageCacheItem and use it in Translation Message. 

The iteration logic her logic is a duplicate of what's already in CompnentService. You used `getMessages(final Locale locale, final String component, boolean useLocaleFallback)` but always passing useLocaleFallback=false, then you iterate here,  which defeats the purpose. 

I will fork your branch and make a code suggestion there.